tasks/task_manager.cc

```diff
@@ -117,48 +179,79 @@ bool task_manager::task::impl::is_done() const noexcept {
     return _status.state == tasks::task_manager::task_state::done;
 }
 
+future<std::vector<task_manager::task::task_essentials>> task_manager::task::impl::get_failed_children() const {
+    return _children.map_each_task<task_essentials>([] (const foreign_task_ptr&) { return std::nullopt; },
+        [] (const task_essentials& child) -> std::optional<task_essentials> {
+            if (child.task_status.state == task_state::failed || !child.failed_children.empty()) {
+                return child;
+            }
+            return std::nullopt;
+        }
+    );
+}
+
 void task_manager::task::impl::run_to_completion() {
-    (void)run().then_wrapped([this] (auto f) {
+    (void)run().then_wrapped([this] (auto f) -> future<> {
         if (f.failed()) {
-            finish_failed(f.get_exception());
+            co_await finish_failed(f.get_exception());
         } else {
+            std::exception_ptr ex = nullptr;
             try {
                 _as.check();
-                finish();
+                co_await finish();
             } catch (...) {
-                finish_failed(std::current_exception());
+                ex = std::current_exception();
+            }
+            if (ex) {
+                co_await finish_failed(ex);
             }
         }
     });
 }
 
-void task_manager::task::impl::finish() noexcept {
+future<> task_manager::task::impl::fold_into_parent() const noexcept {
+    try {
+        if (is_internal() && _parent_id && _children.all_finished()) {
+            auto parent = co_await _module->get_task_manager().lookup_task_on_all_shards(_module->get_task_manager().container(), _parent_id);
+            co_await smp::submit_to(parent.get_owner_shard(), [&parent, id = _status.id] () {
+                return parent->get_children().mark_as_finished(id);
```

Comment:
can we / should we pass the task essentials here instead of calling back into this task from the parent to generate it?

Reply:
If we do that maybe we don't have to copy the failed grandchildren's essentials into the child essentials (the call to get_failed_children) and we can rather move them into the `task_essentials` object before we throw away the task object.  But it's not critical.  We can copy if that makes error handling easier.